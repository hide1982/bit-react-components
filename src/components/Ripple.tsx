import styled, { css } from "styled-components";

interface Styles {
  colors?: string | [string, string?];
  radius?: string;
  opacity?: number;
  duration?: string;
}

const getBorderColor = (colors: Styles["colors"]) => {
  if (!colors) return "rgba(255, 255, 255, 0.3) rgba(0, 0, 0, 0.6)";

  if (typeof colors === "string") {
    return colors;
  }

  return `${colors[0]} ${colors[1] ?? colors[0]}`;
};

interface StyledRippleProps {
  isDisplay: boolean;
  top: number;
  left: number;
  size: number;
  styles?: Styles;
}

const Ripple = styled.span.attrs({ dataTestid: "ripple" })<StyledRippleProps>`
  display: none;
  ${({ isDisplay }) =>
    isDisplay &&
    css`
      display: block;
    `}
  position: absolute;
  top: ${({ top }) => top}px;
  left: ${({ left }) => left}px;
  width: 0;
  height: 0;
  border-radius: ${({ styles }) => styles?.radius ?? 0};
  border-style: solid;
  border-width: ${({ size }) => size / 2}px;
  border-color: ${({ styles }) => getBorderColor(styles?.colors)};
  pointer-events: none;
  opacity: ${({ styles }) => styles?.opacity ?? 0.5};
  transform: scale(0);
  animation: ripple ${({ styles }) => styles?.duration ?? "800ms"} linear;

  @keyframes ripple {
    to {
      transform: scale(4) rotate(45deg);
      opacity: 0;
    }
  }
`;

export default Ripple;
