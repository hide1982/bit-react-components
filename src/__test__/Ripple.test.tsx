import React from "react";
import { render, screen } from "@testing-library/react";
import Ripple from "../components/Ripple";

describe("Ripple component", () => {
  test("snapshot", () => {
    render(<Ripple top={10} left={10} size={10} isDisplay={false} />);
    expect(screen.getByTestId("ripple")).toMatchSnapshot();
  });

  test("Can apply props", () => {
    render(<Ripple top={10} left={10} size={10} isDisplay={false} />);
    expect(screen.getByTestId("ripple")).toHaveStyle({
      top: "10px",
      left: "10px",
      borderWidth: `${10 / 2}px`,
    });
  });
});
