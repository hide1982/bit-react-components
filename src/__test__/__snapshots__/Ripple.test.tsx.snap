// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripple component snapshot 1`] = `
<span
  class="sc-bdfBwQ iwvDkh"
  data-testid="ripple"
  size="10"
/>
`;
